lab 02-04 update cluster (#15)
acend/openshift-4-ops-training: content/en/docs/02/04_update.md

sectionnumber: 2.4
---

> In this lab we will update our OpenShift 4 cluster to a newer minor version.

Reviewer: ```suggestion
In this lab we will update our OpenShift 4 cluster to the latest stable errata release.
```

> Updating an OpenShift 4 cluster is a fully automated process that is managed by several ClusterOperators. If the cluster has Internet access, updates are provided over-the-air, otherwise you need to synchronize the new images to an internal registry prior to starting the update or use a pull-through registry.

Reviewer: ```suggestion
Updating an OpenShift 4 cluster is a fully automated process that is managed by several cluster operators. If the cluster has Internet access, updates are provided over-the-air, otherwise you need to synchronize the new images to an internal registry prior to starting the update or use a pull-through registry.
```

Check ["Updating a restricted network cluster"](https://docs.openshift.com/container-platform/latest/updating/updating-restricted-network-cluster.html) for details on updating a cluster in a disconnected environment.

> The update process continuously rolls out new releases of the ClusterOperators and verifies their readiness before continuing. If a ClusterOperator is stuck in a non ready state, the update does not proceed and manual intervention might be necessary.

Reviewer: ```suggestion
The update process continuously rolls out new releases of all cluster operators and verifies their readiness before continuing. If a cluster operator is stuck in a non-ready state, the update does not proceed and manual intervention might become necessary.
```

> After all relevant ClusterOperators are updated successfully, the Machine Config Operator updates the configuration of the master and worker nodes and - if a newer version is available - replaces the Red Hat Enterprise Linux CoreOS (RHCOS) image of the nodes.

Reviewer: ```suggestion
After all relevant cluster operators are updated successfully, the machine config operator updates the configuration of the master and worker nodes and - if a newer version is available - replaces the nodes' operating system image (Red Hat Enterprise Linux CoreOS).
```

The cluster update can be performed from the web console or the CLI.

## Task {{% param sectionnumber %}}.1a: Updating the cluster from the web console

> Navigate to "Administration" -> "Cluster Settings" to get started. You will see the available update paths in the graph:

Reviewer: ```suggestion
Navigate to **Administration**, then **Cluster Settings** to get started. You will see the available update paths in the graph:
```

![Update overview](resources/update/update-start.png)

To read the release notes of the new version, click on the target version and follow the link:

![Release notes](resources/update/update-release-notes.png)

> The update process from the web console is pretty straight forward. To start the update, all you have to do is press two buttons.

Reviewer: ```suggestion
The update process from the web console is pretty straightforward. To start the update, all you have to do is press two buttons.
```

Click the update button and confirm the update to the target version by again clicking on the update button:

![Start update](resources/update/update-available.png)

> The update process will start updating the ClusterOperators:

Reviewer: ```suggestion
The update process will start updating the cluster operators:
```

![Update started](resources/update/update-update-status-01.png)

> You can see the details of the ClusterOperators in the second tab:

Reviewer: ```suggestion
You can see more details in the second tab:
```

![ClusterOperators status](resources/update/update-co-status.png)

> Worker Nodes may continue to update after the update of Master Nodes and ClusterOperators are complete:

Reviewer: ```suggestion
Worker nodes may continue to update after the update of the masters and cluster operators is complete:
```

![Worker still updating](resources/update/update-worker.png)

Finally, the update is completed:

![Update done](resources/update/update-done.png)

## Task {{% param sectionnumber %}}.1b: Updating the cluster from the CLI

Before starting the update, ensure that your cluster is available:

```bash
oc get clusterversion
```

Example output:

```
NAME      VERSION   AVAILABLE   PROGRESSING   SINCE   STATUS
version   4.7.5     True        False         2d23h   Cluster version is 4.7.5
```

You can check the available updates by running the upgrade command without any parameters:

```bash
oc adm upgrade
```

Example output:

```
Cluster version is 4.7.5

Updates:

VERSION IMAGE
4.7.6   quay.io/openshift-release-dev/ocp-release@sha256:2721145d62b8dbff8808cd7b12a8b24a81a8edad8f1078010f58c673f76bb419
```

{{% alert title="Warning" color="primary" %}}
You can only update to one of the available target versions. Do not force the update to an unsupported version, since it could render your cluster useless.
{{% /alert %}}

> Start the update process:

Reviewer: ```suggestion
Starting the update process depends on what we want to do. We have two options:
```

marketplace                                4.7.5     True        False         False      3d
monitoring                                 4.7.5     True        False         False      74m
network                                    4.7.5     True        False         False      3d
node-tuning                                4.7.5     True        False         False      3d
openshift-apiserver                        4.7.6     True        False         True       3d
openshift-controller-manager               4.7.5     True        False         False      3d
openshift-samples                          4.7.5     True        False         False      3d
operator-lifecycle-manager                 4.7.5     True        False         False      3d
operator-lifecycle-manager-catalog         4.7.5     True        False         False      3d
operator-lifecycle-manager-packageserver   4.7.5     True        False         False      3d
service-ca                                 4.7.5     True        False         False      3d
storage                                    4.7.5     True        False         False      3d
```

{{% alert title="Note" color="primary" %}}

> In the example output above you can see that the ClusterOperator `openshift-apiserver` is already reporting the new version, but is currently in a degraded state. This is expected behaviour during the update process, so don't panic!

Reviewer: ```suggestion
In the example output above you can see that the `openshift-apiserver` ClusterOperator is already reporting the new version, but is currently in a degraded state. This is expected behaviour during the update process, so no need to worry.
```

{{% /alert %}}

> After the ClusterOperators are rolled out, the master and worker nodes are updated. The Machine Config Operator sequentially marks the nodes as unschedulable, applies the new configuration, drains the node and restarts the Kubelet service.

Reviewer: ```suggestion
After all cluster operators are rolled out, the master and worker nodes are updated. The machine config operator sequentially cordons the nodes, applies the new configuration and restarts the Kubelet service.
```
